examples/tensorize_vllm_model.py

```diff
@@ -173,72 +171,94 @@ def deserialize():
-initialize_model_parallel()
+    if args.command == "serialize":
+        eng_args_dict = {f.name: getattr(args, f.name) for f in
+                        dataclasses.fields(EngineArgs)}
 
-keyfile = args.keyfile if args.keyfile else None
-
-
-if args.model_loader_extra_config:
-    config = json.loads(args.model_loader_extra_config)
-    tensorizer_args = TensorizerConfig(**config)._construct_tensorizer_args()
-    tensorizer_args.tensorizer_uri = args.path_to_tensors
-else:
-    tensorizer_args = None
-
-if args.command == "serialize":
-    eng_args_dict = {f.name: getattr(args, f.name) for f in
-                     dataclasses.fields(EngineArgs)}
+        engine_args = EngineArgs.from_cli_args(
+            argparse.Namespace(**eng_args_dict)
+        )
 
-    engine_args = EngineArgs.from_cli_args(argparse.Namespace(**eng_args_dict))
-    engine = LLMEngine.from_engine_args(engine_args)
+        input_dir = args.serialized_directory.rstrip('/')
+        suffix = args.suffix if args.suffix else uuid.uuid4().hex
+        base_path = f"{input_dir}/vllm/{model_ref}/{suffix}"
+        is_sharded = False
+        if engine_args.tensor_parallel_size > 1:
+            model_path = f"{base_path}/model-rank-%03d.tensors"
+            is_sharded = True
+        else:
+            model_path = f"{base_path}/model.tensors"
+            is_sharded = False
+
+        # create and write encryption key before initializing engine to support
+        # sharded models
+        if keyfile is not None:
+            encryption_params = EncryptionParams.random()
+            with _write_stream(keyfile) as stream:
+                stream.write(encryption_params.key)
 
-    input_dir = args.serialized_directory.rstrip('/')
-    suffix = args.suffix if args.suffix else uuid.uuid4().hex
-    base_path = f"{input_dir}/vllm/{model_ref}/{suffix}"
-    model_path = f"{base_path}/model.tensors"
-    tensorizer_config = TensorizerConfig(
-        tensorizer_uri=model_path,
-        **credentials)
-    serialize_vllm_model(engine, tensorizer_config, keyfile)
-elif args.command == "deserialize":
-    if not tensorizer_args:
         tensorizer_config = TensorizerConfig(
-            tensorizer_uri=args.path_to_tensors,
-            encryption_keyfile = keyfile,
-            **credentials
-        )
-    deserialize()
-else:
-    raise ValueError("Either serialize or deserialize must be specified.")
+            tensorizer_uri=model_path,
+            encryption_keyfile=keyfile,
+            is_sharded=is_sharded,
+            **credentials)
+
+        engine = LLMEngine.from_engine_args(engine_args)
+        if args.tensor_parallel_size > 1:
+            engine.model_executor._run_workers("save_tensorized_model",
+                tensorizer_config = tensorizer_config,
+            )
+        else:
+            serialize_vllm_model(
+                engine.model_executor.driver_worker.model_runner.model,
+                tensorizer_config,
```

Comment:
These seemingly represent the multi-GPU and single GPU serializations of the model respectively, although this is a bit unclear given the syntax. It may be worth considering adding some comments or wrapping them in to functions for clarity.